Removing MXIDs from events (#1228)
matrix-org/matrix-doc: proposals/1228-removing-mxids-from-events.md

> # MSC1228: Removing MXIDs from events

Reviewer: Title change suggested, as this is not limited to user identifiers:
```suggestion
# MSC1228: Identifiers revamp
```
Signed-off by: Erkin Alp Güney <erkinalp9035@gmail.com>

Reviewer 2: Might be worth noting that MSC2787 is *supposed* to supersede MSC1228.

## Background

We would like to be able to break the association between a user's ID (such as
`@richvdh:sw1v.org`) and their activity in a room.

The stretch goal is to also remove the association with server names, since for
many users, they are the only user on a server and it is reasonable to be able

> to ask for the removal of any history of `sw1v.org`'s involvement with a room.

Reviewer 3: At now each homeserver have own unique namespace via domain suffix, this prevent from duplicates in federation for user and room id's. But if we replace them to random keys without domain suffix - how we can prevent making duplicated keys by some homeservers (not only collisions, but also manually created dupilcates on some "evil" homeservers)? Homeserver can't check for duplicates in all federated servers before create key, and after creating keys - it will be too late for change it.

Author: You need to sign the create event with the key, so you can only use a given public key as a room ID if you have the private key; in essence, you're asking for someone to break curve25519.

The curve25519 keyspace is 2^256, so for a 50% chance of a collison somewhere within the entire space of matrix rooms, we'd need about 10^38 rooms (see https://en.wikipedia.org/wiki/Birthday_attack). That's a huge number.

Reviewer 3: Thanks for clarifying!

> The general idea presented here is to use a pseudomym in many places where we
> currently use user IDs. The current `@user:server` then becomes a user alias;
> the mapping between alias and the psuedonumous ID is public but can be removed
> in the future.

Reviewer: ```suggestion
The general idea presented here is to use a pseudoynm in many places where we
currently use user IDs. The current `@user:server` then becomes a user alias;
the mapping between alias and the psuedonymous ID is public but can be removed
in the future.
```

User IDs currently appear in the following places in a room:

 * `sender` of each event
 * `state_key` of `m.room.member` events
 * `users` list in `m.room.power_levels` events
 * `creatorUserId` in the content of `m.widget`

Server names appear in the following places:

 * `origin` of each event
 * keys in the `signatures` dict of each event
 * Room IDs
 * Room Aliases
 * `state_keys` in `m.room.aliases` events
 * `matrix.to` permalinks

## Proposal

[This is v3 of this proposal, which in summary is: do proposal v1, but also
introduce a requirement for a global `user_key` at the same time, which in
future will replace mxids as the user's One True Identity. v1 and v2 are
available for reference at
https://docs.google.com/document/d/1ni4LnC_vafX4h4K4sYNpmccS7QeHEFpAcYcbLS-J21Q#heading=h.y1krynr6itl4.]

>  * Each user (currently identified by an mxid) will also have a `user_key`. In

Author: @martindale suggests: Public keys should not be revealed until the first message is sent; use a hashed value instead.

Author: The problem with a hash is that it requires a separate mechanism to distribute the actual key; and I'm not sure how we would control access to that key. If I want to start a conversation with a given user, I need their public user_key: what is to say that I should or should not be permitted it?

   time, this will replace the mxid as your One True Identity; however for now
   they will live in parallel.

   * A `user_key` is represented like `~1:dV3hr3yE9SxhsWEGBJdTho777S8ompkJTh`,
     where `1:` is a version (to allow other systems to be used in future) and

>      the rest is an (unpadded urlsafe-base64ed) ed25519 public key.

Author: @martindale suggests: ed25519 doesn't support hierarchical keys; recommend secp256k1 instead

Author: It's worth noting that we use ed25519 heavily elsewhere in the protocol. Are the advantages of secp256k1 worth the overhead of introducing a separate set of crypto primitives?

>      the rest is an (unpadded urlsafe-base64ed) ed25519 public key.

Reviewer: ```suggestion
     the rest is an (unpadded urlsafe-base64ed) ed448 public key.
```

Reviewer 4: > the number of rooms encountered in a single federation may be greater than 2¹²⁸-1

That isn't really something we need to worry about at this point.  It isn't going to happen in a very, very long time.

Yes, we may want to switch away from ed25519 at some point for various reasons (e.g. something post-quantum), but that would be a change that we do across the board over the whole protocol, and out of scope for this MSC.

>    * Homeservers are responsible for making up keys for their users.

Reviewer: Move the key creation to the client, as that will allow us to apply various constraints upon them for applying a PoW-based antispam on account generation:
```suggestion
   * The client should create the user key at the time of account creation
    and share the decryption key with its homeserver.
```
Signed-off by: Erkin Alp Güney <erkinalp9035@gmail.com>

Reviewer 5: PoW is not a very viable solution today (compute is cheep and phones are slow) and if it becomes viable in the future it will likely require explicit algorithms, not just "generate a key".

If we want some client requirements for signup we should handle these in the signup flow.

Reviewer 6: Not agreeing with the proposal, but PoW is not generally a bad solution. There are variants that are not computationally bound in a way that gives larger machines a significant advantage.

An important point to make PoW work well is an adaptive mechanism that automatically adjusts the toughness of the challenge.

Reviewer 5: But as I said forcing the client to generate they key isn't one of those adaptive mechanisms. Any PoW would be best implemented as an explicit and separate system.

Reviewer 4: Agreed with kevincox that key generation doesn't seem like an appropriate place to put spam prevention mechanisms.  There may be reasons for preferring generating the key on the client, but the given reason doesn't seem sufficient.

Reviewer: @kevincox said:

> But as I said forcing the client to generate they key isn't one of those adaptive mechanisms

You could simply require both decryption and encryption halves of the key to be larger than some *n* (both smaller than some *n* will not work however), which is mathematically difficult in the finite field definition domain of an elliptic curve cryptosystem.

Reviewer 4: This would only provide one specific anti-spam mechanism (computational work).  Doing this elsewhere is trivial and would allow for different methods if required (e.g. using something that is more memory-bound, similar to Argon).  There is simply no benefit to tying it to key generation here.

Reviewer: > using something that is more memory-bound

Memory and I/O bound operations for antispam are out of question as Matrix is designed with thin clients in mind. CPU-bound operations, on the other hand, just takes more time in a client with less resources. Agreed in the part we do not necessarily need to tie it to key generation.

Reviewer: > There may be reasons for preferring generating the key on the client, but the given reason doesn't seem sufficient.

The server creating its own "client key"s and creating a so-called "server user", or the server converting one of the normal client users into a such server user is a viable attack vector in a server-generated user keys situation. With client generated keys, that would be a protocol violation, even if not enforceable over federation.

Reviewer 4: Saying that the keys should be generated on the client rather than on the server has no practical effect on such an attack, as you admit, so that isn't a good enough reason.

   * For now, each homeserver maintains a one-to-one mapping between
     `user_key` and mxid for each of their users. In future, we will look to

>      break this link to allow portability of accounts.

Reviewer 7: A side note: I imagine that older versions of the CS API, for example, would show the mxid in the fields they're in right now to avoid breaking older clients that rely on parsing an ID (to get just the name or server) for whatever reason.

>  * Room IDs also become ed25519 public keys.

Reviewer: ```suggestion
 * Room IDs also become ed448 public keys.
```

>    * They look like: `!Sr_Vj3FIqyQ2WjJ9fWpUXRdz6fX4oFAjKrDmu198PnI`.

Reviewer 8: Why do user keys get to have versions, yet user room IDs and user room keys do not? (Sorry if this doesn't make sense, this is pretty much my first interaction with the spec and I came from one of the blog posts)

>    * They look like: `!Sr_Vj3FIqyQ2WjJ9fWpUXRdz6fX4oFAjKrDmu198PnI`.

Reviewer 8: Why do user keys get to have versions, yet user room IDs and user room keys do not? (Sorry if this doesn't make sense, this is pretty much my first interaction with the spec and I came from one of the blog posts)

Reviewer: User room IDs and keys are specific to that room, hence they are versioned by the room version.

Reviewer: The root key of the user is not, hence that needs to be versioned separately.

   * The server which creates the room is responsible for creating the keypair.

   * The `m.room.create` event is signed with the room id to stop people making
     new rooms which look like old ones. After this point, the private key is
     never needed again. <sup id="a1">[1](#f1)</sup>

>  * Define a `user_room_key`, which is yet another ed25519 public key.

Reviewer: ```suggestion
 * Define a `user_room_key`, which is yet another ed448 public key.
```

   * It looks like `^Noi6WqcDj0QmPxCNQqgezwTlBKrfqehY1u2FyWP9uYw`.

>    * Homeservers are responsible for making up user keys for their users. They

Reviewer 8: ```suggestion
   * Homeservers are responsible for making up user room keys for their users. They
```

     can (and should) use a different key in each room for each user.

   * This `user_room_key` is used where we currently use an mxid in the DAG:
     `sender`, `m.room.member`, `m.room.power_levels`,
     `creatorUserId`. `creator` is removed as per
     [MSC2175](./2175-remove-creator-field.md).

   * Events are **signed by the `user_room_key` of the sender instead of the
     server's key**.

   * If a user leaves and rejoins a room, they should use the same
     `user_room_key` (unless a server admin has manually removed the old
     mapping). This makes ban evasion harder. (It's up to server owners to
     ensure this rule is followed - servers which don't respect it and allow a
     serial abuser to evade bans by issuing different `user_room_keys` are likely
     to suffer whole-server bans.)

 * Invite and join events include:

>    * `mxid_mapping`: field which gives the user's `@user:server` mxid and which

Author: in time, the intention is to replace the mxid with simply a server name.

     must be signed by the server in question, and the signature must be
     verified before the mapping is considered valid.

   * `user_mapping`: contains the `user_key` giving your True Identity, and
     signed by that key. The signature must be verified by receiving
     homeservers for it to be considered a valid invite/join event for a vNext
     room.

### Examples

> `m.room.create`: signed by both the room key and the `user_room_key` of the

Reviewer 9: we haven't defined "room key" at this point - might be better to call it the room ID's key?

   We make an attempt to verify its `mxid_mapping` before persisting it into
   our db. If the sig is wrong, we reject the event at that point.

   If we can't get the key (with a shortish timeout), we handle it as normal
   (and schedule a retry for later).

   Typically we only care about the most recent<sup id="a2">[2](#f2)</sup>
   `mxid_mapping` for each `user_room_key`; when we see another we can cancel any
   pending verification of any previous mapping<sup id="a3">[3](#f3)</sup>. Any
   previously-verified mapping should remain in place until another mapping
   becomes available.

 * When we _backfill_:

   We do the same thing, although in many cases we'll already have active
   mappings for the users in question, so we can ignore any received that
   way. By only honouring the most recent mapping, we gain the correct
   semantics for account portability: authorisation for backfill depends on the
   current location of the user, rather than wherever they were in the past.

 * When we join a _new room_:

   For now we do the same thing (ie, make an attempt to verify the mxid
   mappings in the join events we receive, and time them out quickly). In

>    future we might optimise this so the the mappings are verified lazily.

Reviewer 9: @richvdh how would this work in practice? are you suggesting that you reject events from the DB in retrospect having lazily verified the mapping, and send these to events as redactions of some kind?

This should mean that in the majority of cases, we'll have a verified mxid by
the time we send the event to a client.

For each `user_room_key`, we therefore have:

 * zero or one verified mxid mappings.
 * zero or one incomplete mxid mappings.

> We extend the CS API to include a `verified_sender_mxid` field on any events

Author: @ara4n says: Whilst i think we should have this present by default to aid stupid clients, should we provide the option to filter it out for clients which know they are smart enough to learn mappings from the async mapping bit of the sync response instead?

Author: my reply: maybe? I'm not sure the added complexity is worthwhile just to remove a field from some JSON which you could otherwise ignore.

sent to a client where we know the user's current **verified** mxid.

 * This is included in the interests of helping simple clients do the right
   thing most of the time - but it is annoying and dangerous because it will
   **sometimes** be missing. Still, we don't want to either (a) hold up all
   traffic in the room while we wait for a verification which may never
   succeed; (b) hold back some events while we do a verification for a sender;
   (c) require that all clients always have to wait for an asynchronous
   verification and match them up.

We also add a **new** field to the `/sync` response which tells clients about
mxid mappings as they are resolved.

> Question: should we remove unverified mxid mappings from join events before

Author: Some debate over this at https://docs.google.com/document/d/1ni4LnC_vafX4h4K4sYNpmccS7QeHEFpAcYcbLS-J21Q/edit?disco=AAAAB3zzaso, but everybody apart from me seemed to think we should, and on re-reading, I think I agree.

serving them to the clients, to stop client developers relying on it and
breaking everything?

> 

Reviewer 4: We may need to give details on how to use the user keys to find user devices and device keys in order to do e2e.

### Sending invites

We have a bootstrapping problem for invites in that, until a user joins a room,
we don't know their `user_room_key`.

Also: invite events are supposed to be signed by the invitee, so that other
members of the room can be sure that they have actually received a copy of the
event.

The current invite dance is:

 * inviting server builds a complete invite event, and signs it
 * inviting server sends a copy to invited server, along with some (unsigned)
   state about the room: name, avatar, inviting user's join event
 * invited server checks that request came from server of inviting user
 * invited server also signs the event, and tells the user about it
 * inviting server adds the (double-signed) event to the DAG and sends it to
   the rest of the federation (including the invited server, if it was already
   in the room)

We could change this to:

 * inviting server builds a partial invite event

>  * inviting server PUTs to `/_matrix/federation/v3/invite/<event_id>/<target_mxid>` on invited server

Reviewer 10: Excuse me asking but it is not completely clear to me how the inviting server knows which is the _invited server_? 

Either we invite someone by knowing his/her `user_key` (which isn't associated to any server) or the `mxid` (which is similarly isn't associated to any server, and any similarity between the postfix label and homeserver names are purely coincidental), in any case if the user never have joined any locally known room then we seem not to have any knowledge about his/her whereabouts, of in fact, existence.

Global lookup systems (like DHT) are usually used to look up the location of a cryptographically identified resource in distributed networks (or get a negative response of non-existence after a timeout) but as far as I know we don't have that.

(I feel a slight similarity to the `room_id` problem, eg. joining a room by not using an alias, where we're supposed to know or figure out one or more originating servers.)

Reviewer 4: My understanding is that we are still going to have mxids of the form `@ localpart:servername`, and if you move servers, then you would get a new `servername` (and possible different `localpart`), and all your room mappings would be updated.  So the `servername` part of your mxid will always be the server that you're contacted at.  If you've moved servers, then I guess your old server should return an error code if someone tries to talk to you (maybe even including a forwarding address).
